Add `Faculty` class

JerryWang0000 opened the pull request:
Add Faculty and its test class.

closes #45

JerryWang0000 committed:
Merge branch 'master' of https://github.com/JerryWang0000/tp

* 'master' of https://github.com/JerryWang0000/tp:
  Update file names to follow format
  Update aboutus page with my details and photo
  Update typos
  Change use case for filter and typo in sort
  aboutus update file name
  changed names to match github names
  update aboutus info to name match lowercase requirements
  Updated use cases
  Update AboutUs for Jerry Wang

JerryWang0000 committed:
Merge branch 'AY2324S2-CS2103-F08-3:master' into master

JerryWang0000 committed:
Merge branch 'master' of https://github.com/JerryWang0000/tp

* 'master' of https://github.com/JerryWang0000/tp:
  Update files with traces of the package name "seedu.address" to "staffconnect"
  Update remaining files with traces of seedu.address in test
  Refactor package seedu.address to staffconnect in test
  Refactor package seedu.address to staffconnect in main
  Update header in _base.scss
  Fix typos and added clarity
  Update UserGuide description
  Update website main index
  Add pluiexo.md page
  Amend Readme
  Update github info
  Update Readme to fit StaffConnect
  Update DeveloperGuide.md
  Update AboutUs: Replace placeholder images
  Update UserGuide.md

JerryWang0000 committed:
Merge branch 'AY2324S2-CS2103-F08-3:master' into master

JerryWang0000 committed:
Merge branch 'master' of https://github.com/JerryWang0000/tp

JerryWang0000 committed:
Add Faculty class

Add Faculty and its test class.

JerryWang0000 committed:
Add JavaDoc

JerryWang0000 committed:
Make minor changes

tsulim commented on src/main/java/staffconnect/model/person/Faculty.java:
```diff
@@ -0,0 +1,52 @@
+            return ((Faculty) obj).value.equals(this.value);
+        }
+        return false;
+    }
```
Perhaps you can consider [making the happy path prominent](https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent)? You can also refer to `name.java` if you need a reference.

tsulim commented on src/main/java/staffconnect/model/person/Faculty.java:
```diff
@@ -0,0 +1,52 @@
+     * The first character of the faculty must not be a whitespace,
+     * otherwise " " (a blank string) becomes a valid input.
+     */
+    public static final String VALIDATION_REGEX = "[^\\s].*";
```
I noticed that you're accepting non-alphanumeric characters. Could you please provide some insight into the rationale behind this decision? It might be beneficial to clarify the intention for handling such characters.

iynixil replied:
I agree, perhaps the values for "Faculty" could be limited to schools in NUS as we have specified our product to be for NUS students?

whitesnowx replied:
If we are limiting to NUS schools, should it be an enum?

iynixil replied:
I believe so, that would mean we can save standardized faculty values in our app if we wish to do so. e.g user enters "School of Computing" -> saved as "SOC" or vice versa.

JerryWang0000 replied:
It is possible but should take a bit long to implement

tsulim reviewed (commented):
Overall, it looks good to me! However, I have a couple of suggestions written in the review.
Keep up the good work, and let me know if you have any questions or need further clarification on the suggestions! 👍

iynixil reviewed (commented):
Looks good! Could do with a few improvements (the equals() method, regex pattern).
Left a suggestion for the values that Faculty can take.

JerryWang0000 committed:
Modify the euqals() method

JerryWang0000 committed:
Modify Faculty to use enums

Restrict the valid input values to currently existing NUS faculties. Change the test class.

JerryWang0000 committed:
Make minor changes

JerryWang0000 commented:
Now I store the faculty value as an enum. Do I need to change it to String?

JerryWang0000 committed:
Change Type of value in Faculty.java

Change the type of value to String.

Pluiexo commented on src/main/java/staffconnect/model/person/Faculty.java:
```diff
@@ -0,0 +1,113 @@
+    public Faculty(String faculty) {
+        requireNonNull(faculty);
+        checkArgument(isValidFaculty(faculty), MESSAGE_CONSTRAINTS);
+        value = fromString(faculty).getFacultyName(); // can be extended
```
So far so good, though the logic is abit off currenly it is: string -> enum->string, it makes having the enum class a bit redundant. May I suggest storing the enum instead and returning the enum. This gives an additional layer of abstraction for those who use faculty.

JerryWang0000 replied:
I did consider that, but I wish to keep conformity as other attributes which are also strings. The fromString() function can be extended so that maybe syncronyms can be allowed. don't have to type in the full name of the faculty. Adding enum can simplify the process.

Pluiexo commented on src/main/java/staffconnect/model/person/Faculty.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Override
+    public String toString() {
+        return value;
```
When storing value as a enum, you may choose call a toString in the enum class to return the value inside the enums instead

Pluiexo commented on src/main/java/staffconnect/model/person/Faculty.java:
```diff
@@ -0,0 +1,113 @@
+     * For this version, a valid faculty value should match exactly the full name of the facult,
+     * or the value is invalid
+     */
+    public enum FacultyEnum {
```
For conciseness Enums don't contain the term enums in them, rather they are nouns by themselves to represent the type of items they contain. Eg: Faculties

Pluiexo reviewed (commented):
So far so good! Though several logical nits to bring up

JerryWang0000 committed:
Fix minor typos in JavaDoc

JerryWang0000 committed:
Change value type of Faculty to Enum

JerryWang0000 committed:
Rename the name of FacultyEnum

Rename the enum to FacultyName

Pluiexo commented on src/main/java/staffconnect/model/person/Faculty.java:
```diff
@@ -0,0 +1,113 @@
+     * @return True if the input matches any faculty name.
+     */
+    public static boolean isValidFaculty(String test) {
+        requireNonNull(test, "faculty cannot be null");
```
Some may consider this magic string, but you don't have to fix it in this commit

Pluiexo reviewed (approved):
好嘞
